### PR TITLE
optimizer: implements Join

### PR DESCRIPTION
**Pull request by coocood**
Support Join in new plan.

In this PR, optimization information is collected, but not used, inner joins are joined in the original order.

Futher optimization and tests will be supported in the next PR.

### CODE REVIEW

File: optimizer/plan/planbuilder_join.go [hunk 1]
```diff
@@ -0,0 +1,710 @@
...
+
+import (
+	"github.com/ngaut/log"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/parser/opcode"
+)
+
+// Equiv represents a equivalent join condition.
```
Comment (shenli) [inline optimizer/plan/planbuilder_join.go]: a->an

File: optimizer/plan/planbuilder_join.go [hunk 2]
```diff
@@ -0,0 +1,710 @@
...
+// newTablePath creates a new table join path.
+func newTablePath(table *ast.TableName) *joinPath {
+	return &joinPath{
+		table:      table,
+		filterRate: 1,
+	}
+}
+
+// newOuterJoinPath creates a new outer join path and pushes on condition to children paths.
+// the returned joinPath slice has one element.
```
Comment (shenli) [inline optimizer/plan/planbuilder_join.go]: the-> The

File: optimizer/plan/planbuilder_join.go [hunk 3]
```diff
@@ -0,0 +1,710 @@
...
+}
+
+// joinPath can be a single table path, inner join or outer join.
+type joinPath struct {
+	// for table path
+	table           *ast.TableName
+	totalFilterRate float64
+
+	neighborCount int // number of neighbor table.
+	idxDepCount   int // number of tables this table depends on.
```
Comment (shenli) [inline optimizer/plan/planbuilder_join.go]: tables->joinPaths?

File: executor/executor.go [hunk 4]
```diff
@@ -525,6 +525,147 @@ func (e *IndexScanExec) Close() error {
 	return nil
 }
 
+// JoinOuterExec represents a outer join executor.
```
Comment (shenli) [inline executor/executor.go]: a->an

File: executor/executor.go [hunk 5]
```diff
@@ -525,6 +525,147 @@ func (e *IndexScanExec) Close() error {
...
+	gotRow    bool
+}
+
+// Fields implements Executor Fields interface.
+func (e *JoinOuterExec) Fields() []*ast.ResultField {
+	return e.fields
+}
+
+// Next implements Executor Next interface.
+// If inner executor didn't get any row for a outer executor row,
```
Comment (shenli) [inline executor/executor.go]: a->an

File: executor/executor.go [hunk 6]
```diff
@@ -525,6 +525,147 @@ func (e *IndexScanExec) Close() error {
...
+		}
+		var match = true
+		if e.Condition != nil {
+			match, err = evaluator.EvalBool(e.ctx, e.Condition)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+		}
+		if match {
+			return row, nil
```
Comment (shenli) [inline executor/executor.go]: Add some comment here. The row's value is not used by the outer executor.

File: executor/executor.go [hunk 7]
```diff
@@ -525,6 +525,147 @@ func (e *IndexScanExec) Close() error {
...
+
+// Next implements Executor Next interface.
+func (e *JoinInnerExec) Next() (*Row, error) {
+	if e.done {
+		return nil, nil
+	}
+	for {
+		exec := e.innerExecs[e.cursor]
+		if exec == nil {
+			exec = e.builder.build(e.InnerPlans[e.cursor])
```
Comment (shenli) [inline executor/executor.go]: Call plan.Refine before build()?

Comment (coocood) [inline executor/executor.go]: You're right.

File: executor/builder.go [hunk 8]
```diff
@@ -162,7 +166,32 @@ func (b *executorBuilder) buildIndexRange(scan *IndexScanExec, v *plan.IndexRang
...
+		Condition:  b.joinConditions(v.Conditions),
+		fields:     v.Fields(),
+		ctx:        b.ctx,
+		builder:    b,
+	}
+	return e
+}
+
 func (b *executorBuilder) joinConditions(conditions []ast.ExprNode) ast.ExprNode {
+	if len(conditions) == 0 {
```
Comment (c4pt0r) [inline executor/builder.go]: useless code

if len(condition) == 1 {
    return condition[0]
}
return nil

File: executor/builder.go [hunk 9]
```diff
@@ -57,10 +57,14 @@ func (b *executorBuilder) build(p plan.Plan) Executor {
 		return b.buildHaving(v)
 	case *plan.IndexScan:
 		return b.buildIndexScan(v)
+	case *plan.JoinOuter:
+		return b.buildJoinOuter(v)
 	case *plan.Limit:
 		return b.buildLimit(v)
 	case *plan.Prepare:
 		return b.buildPrepare(v)
+	case *plan.JoinInner:
```
Comment (qiuyesuifeng) [inline executor/builder.go]: It is better to sort it.

File: executor/builder.go [hunk 10]
```diff
@@ -162,7 +166,32 @@ func (b *executorBuilder) buildIndexRange(scan *IndexScanExec, v *plan.IndexRang
 	return ran
 }
 
+func (b *executorBuilder) buildJoinOuter(v *plan.JoinOuter) *JoinOuterExec {
+	e := &JoinOuterExec{
+		OuterExec: b.build(v.Outer),
+		InnerPlan: v.Inner,
+		fields:    v.Fields(),
+		builder:   b,
```
Comment (qiuyesuifeng) [inline executor/builder.go]: Why outer join has no Condition member?

File: executor/executor.go [hunk 11]
```diff
@@ -525,6 +525,150 @@ func (e *IndexScanExec) Close() error {
...
+// Close implements Executor Close interface.
+func (e *JoinOuterExec) Close() error {
+	err := e.OuterExec.Close()
+	if e.innerExec != nil {
+		return e.innerExec.Close()
+	}
+	return err
+}
+
+// JoinInnerExec represents a inner join executor.
```
Comment (qiuyesuifeng) [inline executor/executor.go]: an inner join

File: executor/executor.go [hunk 12]
```diff
@@ -525,6 +525,150 @@ func (e *IndexScanExec) Close() error {
...
+	}
+}
+
+// Close implements Executor Close interface.
+func (e *JoinOuterExec) Close() error {
+	err := e.OuterExec.Close()
+	if e.innerExec != nil {
+		return e.innerExec.Close()
+	}
+	return err
```
Comment (qiuyesuifeng) [inline executor/executor.go]: errors.Trace(err)

File: executor/executor.go [hunk 13]
```diff
@@ -525,6 +525,150 @@ func (e *IndexScanExec) Close() error {
...
+// The data in the returned row is not used by caller.
+func (e *JoinInnerExec) Next() (*Row, error) {
+	if e.done {
+		return nil, nil
+	}
+	for {
+		exec := e.innerExecs[e.cursor]
+		if exec == nil {
+			plan.Refine(e.InnerPlans[e.cursor])
+			exec = e.builder.build(e.InnerPlans[e.cursor])
```
Comment (qiuyesuifeng) [inline executor/executor.go]: Define `e.InnerPlans[e.cursor] ` as a variable is better.

File: executor/executor.go [hunk 14]
```diff
@@ -525,6 +525,150 @@ func (e *IndexScanExec) Close() error {
...
+			return row, nil
+		}
+	}
+}
+
+// Close implements Executor Close interface.
+func (e *JoinInnerExec) Close() error {
+	for _, inExec := range e.innerExecs {
+		if inExec != nil {
+			inExec.Close()
```
Comment (qiuyesuifeng) [inline executor/executor.go]: Should we handle return error?

File: executor/executor.go [hunk 15]
```diff
@@ -525,6 +525,150 @@ func (e *IndexScanExec) Close() error {
...
+
+// Fields implements Executor Fields interface.
+func (e *JoinOuterExec) Fields() []*ast.ResultField {
+	return e.fields
+}
+
+// Next implements Executor Next interface.
+// The data in the returned row is not used by caller.
+// If inner executor didn't get any row for an outer executor row,
+// A row with 0 len Data indicates there is no inner row matched for
```
Comment (qiuyesuifeng) [inline executor/executor.go]: a row

File: optimizer/plan/planbuilder_join.go [hunk 16]
```diff
@@ -0,0 +1,710 @@
...
+import (
+	"github.com/ngaut/log"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/parser/opcode"
+)
+
+// Equiv represents an equivalent join condition, like "t1.c1 = t2.c1".
+type Equiv struct {
```
Comment (qiuyesuifeng) [inline optimizer/plan/planbuilder_join.go]: How about `EqCondition` or `EqualCond`?

Comment (coocood) [inline optimizer/plan/planbuilder_join.go]: Alright, but not in this PR, or there would be much confliction.

File: optimizer/plan/planbuilder_join.go [hunk 17]
```diff
@@ -0,0 +1,710 @@
...
+	eq.RightIdx = equivHasIndex(eq.Right)
+	return eq
+}
+
+func equivHasIndex(rf *ast.ResultField) bool {
+	if rf.Table.PKIsHandle && mysql.HasPriKeyFlag(rf.Column.Flag) {
+		return true
+	}
+	for _, idx := range rf.Table.Indices {
+		if len(idx.Columns) == 1 && idx.Columns[0].Name.L == rf.Column.Name.L {
```
Comment (qiuyesuifeng) [inline optimizer/plan/planbuilder_join.go]: Is there any need to extract a function? 
Seems many places have used this.

File: optimizer/plan/planbuilder_join.go [hunk 18]
```diff
@@ -0,0 +1,710 @@
...
+	// The joinPaths that this path's index depends on.
+	idxDeps   map[*joinPath]bool
+	neighbors map[*joinPath]bool
+}
+
+// newTablePath creates a new table join path.
+func newTablePath(table *ast.TableName) *joinPath {
+	return &joinPath{
+		table:      table,
+		filterRate: 1,
```
Comment (qiuyesuifeng) [inline optimizer/plan/planbuilder_join.go]: Use a const variable.

File: executor/builder.go (hunk 10 above)
Comment (coocood) [inline executor/builder.go]: All conditions should be pushed to table.

File: optimizer/plan/planbuilder_join.go (hunk 17 above)
Comment (coocood) [inline optimizer/plan/planbuilder_join.go]: Just two places, and one of them will be removed later.

File: optimizer/plan/planbuilder_join.go [hunk 19]
```diff
@@ -0,0 +1,710 @@
...
+		if isOuter {
+			return newOuterJoinPath(x.Tp == ast.RightJoin, leftPath, righPath, x.On)
+		}
+		return newInnerJoinPath(leftPath, righPath, x.On)
+	case *ast.TableSource:
+		return b.buildBasicJoinPath(x.Source, nullRejectTables)
+	case *ast.TableName:
+		return newTablePath(x)
+	default:
+		b.err = ErrUnsupportedType.Gen("unsupported table source type %")
```
Comment (qiuyesuifeng) [inline optimizer/plan/planbuilder_join.go]: Seems uncompleted?

File: optimizer/plan/planbuilder_join.go [hunk 20]
```diff
@@ -0,0 +1,710 @@
...
+	return newEquiv(ln.Refer, rn.Refer)
+}
+
+// pathMap is the inner paths
+func (b *planBuilder) nextPath(pathMap map[*ast.TableName]*joinPath,
+	equivMap map[*Equiv]bool, prevs []*joinPath) *joinPath {
+	cans := b.candidatePaths(pathMap)
+	if len(cans) == 0 {
+		for _, v := range pathMap {
+			log.Errorf("index dep %v, prevs %v\n", v.idxDeps, len(prevs))
```
Comment (qiuyesuifeng) [inline optimizer/plan/planbuilder_join.go]: Remove this log?

Comment (coocood) [inline optimizer/plan/planbuilder_join.go]: This code is not used right now.

**Comment by shenli** [comment]
LGTM

**Comment by qiuyesuifeng** [comment]
LGTM

**Pull request merged**